File: frontend/src/keycloak.js
import Keycloak from 'keycloak-js'
import { LOGIN_CHANNEL } from './login-callback.js'

/**
 * Signing in, keeping the session alive, and signing out.
 *
 * These three values must match the realm import exactly:
 *   url      -> where Keycloak is running
 *   realm    -> platform
 *   clientId -> platform-web, a PUBLIC client
 *
 * A browser app cannot keep a client secret — anyone can read the JavaScript —
 * so platform-web is public. Its security comes from two other places: the
 * registered redirect URIs, and PKCE.
 *
 * Note what this app never sees: the OneID client secret, the OneID access
 * token, and the user's PIN. Only Keycloak talks to OneID, and only Keycloak
 * holds the link to the national identity. The browser holds a Keycloak token
 * and nothing else.
 */
const config = {
  url: import.meta.env.VITE_KEYCLOAK_URL ?? 'http://localhost:8190',
  realm: import.meta.env.VITE_KEYCLOAK_REALM ?? 'platform',
  clientId: import.meta.env.VITE_KEYCLOAK_CLIENT_ID ?? 'platform-web'
}
const realmUrl = `${config.url}/realms/${config.realm}`

/** The popup's last stop, registered under platform-web's redirect URIs. */
const callbackUrl = () => `${window.location.origin}/auth-callback.html`

/** How often to notice that the person closed the sign-in window. */
const POPUP_POLL_MS = 500
/** How long a closed window may still deliver its answer, which is posted just before it closes. */
const POPUP_CLOSED_GRACE_MS = 1000

export class LoginError extends Error {
  constructor(code, message) {
    super(message)
    this.name = 'LoginError'
    this.code = code
  }
}

/** The keycloak-js instance of the signed-in session, or null when signed out. */
let current = null
/** The sign-in in progress, so a second click brings its window forward instead of opening another. */
let pending = null

const sessionEndedListeners = new Set()

/**
 * Runs whenever this tab stops being signed in: Logout here, Logout in another
 * tab, or a token refresh Keycloak refused because the session is gone.
 */
export function onSessionEnded(listener) {
  sessionEndedListeners.add(listener)
}

function sessionEnded() {
  current = null
  sessionEndedListeners.forEach((listener) => listener())
}

/**
 * Tells the other tabs of this app that the person signed out.
 *
 * Tokens live in memory, per tab. Without this, a second tab would carry on
 * showing a signed-in page until its next token refresh was refused.
 */
const authChannel = typeof BroadcastChannel === 'undefined' ? null : new BroadcastChannel('platform-auth')

authChannel?.addEventListener('message', (event) => {
  if (event.data === 'logout') {
    current?.clearToken()
  }
})

/**
 * Signs in through a separate window.
 *
 *   1. A popup opens on Keycloak's authorization endpoint. With the default
 *      idpHint "oneid", Keycloak skips its own sign-in page and sends the popup
 *      straight to OneID.
 *   2. The person signs in at OneID, inside the popup.
 *   3. OneID returns the popup to Keycloak. Keycloak, on its server, exchanges
 *      OneID's code for OneID's token, fetches the person's data with it, creates
 *      or updates the Keycloak user, and then sends the popup to
 *      /auth-callback.html with a Keycloak code.
 *   4. The callback page hands that code to this window and closes. This window
 *      redeems it for Keycloak's tokens, with the PKCE verifier only it holds.
 *
 * The OneID token never reaches the browser: steps 3's exchange and lookup
 * happen between Keycloak and OneID only.
 *
 * Pass idpHint: null for Keycloak's own sign-in page, where the development users
 * sign in with a password.
 *
 * NOT async on purpose: the popup must open in the same tick as the click. After
 * any await, browsers treat window.open as unsolicited and block it.
 */
export function login({ idpHint = 'oneid' } = {}) {
  if (pending) {
    pending.popup.focus?.()
    return pending.promise
  }

  const popup = window.open('', 'platform-login', popupFeatures())
  if (!popup) {
    return Promise.reject(new LoginError('popup_blocked',
      'The browser blocked the sign-in window. Allow pop-ups for this site and try again.'))
  }

  const promise = signInThroughPopup(popup, idpHint).finally(() => {
    pending = null
    if (!popup.closed) {
      popup.close()
    }
  })
  pending = { popup, promise }
  return promise
}

async function signInThroughPopup(popup, idpHint) {
  // state ties the answer to this attempt (the CSRF defence for the redirect),
  // nonce ties the ID token to it, and the PKCE verifier makes the code useless to
  // anyone who sees it. All three stay in this function's memory.
  const state = randomString()
  const nonce = randomString()
  const verifier = randomString()
  const challenge = await sha256Base64Url(verifier)

  // Listen before navigating, so no answer can arrive unheard.
  const answer = waitForCallback(popup, state)
  popup.location.href = authorizationUrl({ state, nonce, challenge, idpHint })
  const { code, iss } = await answer

  // RFC 9207: an answer claiming to come from another issuer is not ours to redeem.
  if (iss && iss !== realmUrl) {
    throw new LoginError('issuer_mismatch', 'The sign-in answer came from an unexpected server.')
  }

  const tokens = await exchangeCode(code, verifier)
  if (claims(tokens.id_token).nonce !== nonce) {
    throw new LoginError('nonce_mismatch', 'The sign-in answer does not belong to this sign-in attempt.')
  }

  // A fresh keycloak-js instance per session: an instance can be initialised only
  // once, and it takes over from here — token refresh, logout URL, clearing tokens.
  const keycloak = new Keycloak(config)
  await keycloak.init({
    token: tokens.access_token,
    refreshToken: tokens.refresh_token,
    idToken: tokens.id_token,
    pkceMethod: 'S256',
    // The hidden-iframe session check needs third-party cookies and only adds
    // confusion in a learning project. Off on purpose.
    checkLoginIframe: false
  })
  // keycloak-js fires this from clearToken(), which every way out goes through.
  keycloak.onAuthLogout = sessionEnded
  current = keycloak
}

/** Resolves with the callback page's answer for this attempt, or rejects when the window is closed first. */
function waitForCallback(popup, state) {
  return new Promise((resolve, reject) => {
    const channel = new BroadcastChannel(LOGIN_CHANNEL)
    let done = false
    let closedSeen = false

    const finish = (settle, value) => {
      if (done) return
      done = true
      clearInterval(watch)
      channel.close()
      settle(value)
    }

    channel.addEventListener('message', (event) => {
      const message = event.data
      // Another tab's sign-in, or a stale one: not this attempt, keep waiting.
      if (!message || message.state !== state) return
      if (message.error) {
        finish(reject, new LoginError(message.error, message.errorDescription || 'Signing in was not completed.'))
      } else if (!message.code) {
        finish(reject, new LoginError('no_code', 'The sign-in window returned without an authorization code.'))
      } else {
        finish(resolve, message)
      }
    })

    const watch = setInterval(() => {
      if (popup.closed && !closedSeen) {
        closedSeen = true
        setTimeout(() => finish(reject, new LoginError('popup_closed',
          'The sign-in window was closed before signing in finished.')), POPUP_CLOSED_GRACE_MS)
      }
    }, POPUP_POLL_MS)
  })
}

function authorizationUrl({ state, nonce, challenge, idpHint }) {
  const params = new URLSearchParams({
    client_id: config.clientId,
    redirect_uri: callbackUrl(),
    response_type: 'code',
    response_mode: 'query',
    scope: 'openid',
    state,
    nonce,
    code_challenge: challenge,
    code_challenge_method: 'S256'
  })
  if (idpHint) {
    params.set('kc_idp_hint', idpHint)
  }
  return `${realmUrl}/protocol/openid-connect/auth?${params}`
}

async function exchangeCode(code, verifier) {
  let response
  try {
    response = await fetch(`${realmUrl}/protocol/openid-connect/token`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: new URLSearchParams({
        grant_type: 'authorization_code',
        client_id: config.clientId,
        code,
        redirect_uri: callbackUrl(),
        code_verifier: verifier
      })
    })
  } catch {
    throw new LoginError('network', 'Keycloak could not be reached to finish signing in.')
  }
  const body = await response.json().catch(() => ({}))
  if (!response.ok || !body.access_token) {
    throw new LoginError(body.error || 'token_exchange_failed', body.error_description || 'Keycloak refused to finish signing in.')
  }
  return body
}

/**
 * Ends the Keycloak session, not just the local tokens.
 *
 * Clearing tokens in the browser is not logout, it only hides them: the Keycloak
 * session would still be live and the next login would silently walk straight
 * back in. So the browser is sent to Keycloak's end-session endpoint, which
 * deletes the session server-side and revokes its refresh token.
 *
 * The order matters. The URL is built first because it carries the ID token as
 * id_token_hint; with that hint and a registered post_logout_redirect_uri,
 * Keycloak ends the session without asking "Do you want to log out?". Only then
 * are the tokens forgotten, in this tab and the others, so nothing that still
 * looks signed in survives even if the redirect itself fails.
 *
 * Note also what this does NOT do by default — it does not end the OneID session
 * at sso.egov.uz. That is Keycloak's job, behind ONEID_CALL_LOGOUT, never the
 * browser's: the browser has no OneID token to end it with.
 */
export function logout() {
  const keycloak = current
  if (!keycloak) {
    return
  }
  const logoutUrl = keycloak.createLogoutUrl({ redirectUri: window.location.origin + '/' })
  authChannel?.postMessage('logout')
  keycloak.clearToken()
  // replace, not assign: Back must not return to a page that was signed in.
  window.location.replace(logoutUrl)
}

/**
 * A currently valid access token, refreshed if it expires within 30 seconds.
 *
 * Access tokens live 5 minutes in this realm. Without this the app would start
 * collecting 401s after five minutes of use.
 */
export async function getValidToken() {
  const keycloak = current
  if (!keycloak?.authenticated) {
    return null
  }
  try {
    await keycloak.updateToken(30)
  } catch (error) {
    // Two different failures land here.
    //
    // Keycloak answered 400: the session is gone — logged out in another tab,
    // ended by an administrator, or idle past the realm's timeout. keycloak-js
    // has already cleared its tokens, which fired onSessionEnded. Deliberately no
    // automatic sign-in here: it would open a window in the middle of a click,
    // when what the person needs to see is that they were signed out.
    //
    // Keycloak unreachable: the tokens are kept, and the current one is sent as
    // it is. The backend then decides, which is the honest answer either way.
    console.warn('Token refresh failed', error)
  }
  return keycloak.authenticated ? keycloak.token : null
}

export function isAuthenticated() {
  return current?.authenticated === true
}

/** Realm roles as Keycloak decoded them from the access token. */
export function realmRoles() {
  return current?.realmAccess?.roles ?? []
}

export function username() {
  return current?.tokenParsed?.preferred_username ?? '-'
}

// ---- small helpers ------------------------------------------------------------

function popupFeatures() {
  const width = 520
  const height = 720
  const left = Math.max(0, Math.round((window.screenX || 0) + ((window.outerWidth || width) - width) / 2))
  const top = Math.max(0, Math.round((window.screenY || 0) + ((window.outerHeight || height) - height) / 2))
  return `popup=yes,width=${width},height=${height},left=${left},top=${top}`
}

function randomString() {
  return base64Url(crypto.getRandomValues(new Uint8Array(32)))
}

async function sha256Base64Url(value) {
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(value))
  return base64Url(new Uint8Array(digest))
}

function base64Url(bytes) {
  return btoa(String.fromCharCode(...bytes)).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '')
}

/** A token's claims, read only to compare the nonce. Its signature is Keycloak's to vouch for, over TLS. */
function claims(jwt) {
  const base64 = jwt.split('.')[1].replace(/-/g, '+').replace(/_/g, '/')
  const bytes = Uint8Array.from(atob(base64), (character) => character.charCodeAt(0))
  return JSON.parse(new TextDecoder().decode(bytes))
}

File: e2e-tests/src/test/java/uz/platform/e2e/support/Keycloak.java
package uz.platform.e2e.support;

import static java.nio.charset.StandardCharsets.UTF_8;
import static uz.platform.e2e.support.Platform.APP;
import static uz.platform.e2e.support.Platform.DEV_USER_PASSWORD;
import static uz.platform.e2e.support.Platform.KEYCLOAK;
import static uz.platform.e2e.support.Platform.LOGIN_CALLBACK;
import static uz.platform.e2e.support.Platform.REALM;
import static uz.platform.e2e.support.Platform.secret;

import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.UUID;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import uz.platform.e2e.support.Http.Response;

/** Keycloak's public endpoints as a client uses them, and its admin API for inspection. */
public final class Keycloak {

    private static final String TOKEN = REALM + "/protocol/openid-connect/token";

    public record Tokens(String accessToken, String refreshToken, String idToken) {

        public JsonNode claims() {
            return Jwts.claims(accessToken);
        }

        public String subject() {
            return claims().path("sub").asText();
        }

        public String sessionId() {
            return claims().path("sid").asText();
        }

        public long expiresAt() {
            return claims().path("exp").asLong();
        }
    }

    private Keycloak() {
    }

    // ---- as platform-web ----------------------------------------------------

    /** A development user's tokens by password grant, which platform-web allows for scripted checks. */
    public static Tokens password(String username) {
        Response response = Http.post(TOKEN).form(Map.of(
                "grant_type", "password",
                "client_id", "platform-web",
                "scope", "openid",
                "username", username,
                "password", DEV_USER_PASSWORD)).send();
        return tokens(expect(response, 200, "password grant for " + username));
    }

    /** The authorization-code exchange the app performs after the popup, PKCE verifier included. */
    public static Tokens exchangeCode(String code, String codeVerifier) {
        Response response = Http.post(TOKEN).form(Map.of(
                "grant_type", "authorization_code",
                "client_id", "platform-web",
                "code", code,
                "redirect_uri", LOGIN_CALLBACK,
                "code_verifier", codeVerifier)).send();
        return tokens(expect(response, 200, "authorization code exchange"));
    }

    public static Response refresh(String refreshToken) {
        return Http.post(TOKEN).form(Map.of(
                "grant_type", "refresh_token",
                "client_id", "platform-web",
                "refresh_token", refreshToken)).send();
    }

    public static Response userinfo(String accessToken) {
        return Http.get(REALM + "/protocol/openid-connect/userinfo").bearer(accessToken).send();
    }

    /**
     * The URL the app's sign-in popup opens. {@code kc_idp_hint=oneid} makes Keycloak skip
     * its own sign-in page and go straight to OneID.
     */
    public static String authorizationUrl(String codeChallenge) {
        return REALM + "/protocol/openid-connect/auth?" + Http.formEncode(Map.of(
                "client_id", "platform-web",
                "redirect_uri", LOGIN_CALLBACK,
                "response_type", "code",
                "response_mode", "query",
                "scope", "openid",
                "state", UUID.randomUUID().toString(),
                "code_challenge", codeChallenge,
                "code_challenge_method", "S256",
                "kc_idp_hint", "oneid"));
    }

    /** The end-session URL exactly as keycloak-js builds it for logout(). */
    public static String endSessionUrl(String idToken) {
        Map<String, String> query = new LinkedHashMap<>();
        query.put("client_id", "platform-web");
        query.put("post_logout_redirect_uri", APP);
        query.put("id_token_hint", idToken);
        return REALM + "/protocol/openid-connect/logout?" + Http.formEncode(query);
    }

    public static JsonNode jwks() {
        return expect(Http.get(REALM + "/protocol/openid-connect/certs").send(), 200, "JWKS");
    }

    // ---- as a confidential service client -----------------------------------

    public static Response clientCredentials(String clientId, String clientSecret) {
        return Http.post(TOKEN).form(Map.of(
                "grant_type", "client_credentials",
                "client_id", clientId,
                "client_secret", clientSecret)).send();
    }

    /** A service's access token, with its secret from USER_SERVICE_CLIENT_SECRET and the like. */
    public static String serviceToken(String clientId) {
        String secret = secret(clientId.toUpperCase(Locale.ROOT).replace('-', '_') + "_CLIENT_SECRET");
        return expect(clientCredentials(clientId, secret), 200, "client_credentials for " + clientId)
                .path("access_token").asText();
    }

    /** Keycloak's own view of a token, asked as user-service, which is a confidential client. */
    public static boolean isActive(String token) {
        String credentials = Base64.getEncoder()
                .encodeToString(("user-service:" + secret("USER_SERVICE_CLIENT_SECRET")).getBytes(UTF_8));
        Response response = Http.post(TOKEN + "/introspect")
                .header("Authorization", "Basic " + credentials)
                .form(Map.of("token", token))
                .send();
        return expect(response, 200, "token introspection").path("active").asBoolean();
    }

    // ---- admin API, for inspecting what Keycloak stored ----------------------

    /** A real Keycloak token, from the master realm: a different issuer than every service trusts. */
    public static String masterRealmToken() {
        Response response = Http.post(KEYCLOAK + "/realms/master/protocol/openid-connect/token").form(Map.of(
                "grant_type", "password",
                "client_id", "admin-cli",
                "username", secret("KEYCLOAK_ADMIN_USER"),
                "password", secret("KEYCLOAK_ADMIN_PASSWORD"))).send();
        return expect(response, 200, "master realm admin token").path("access_token").asText();
    }

    public static JsonNode admin(String path) {
        return expect(Http.get(adminUrl(path)).bearer(masterRealmToken()).send(), 200, "GET admin" + path);
    }

    public static JsonNode user(String username) {
        JsonNode found = admin("/users?exact=true&username=" + URLEncoder.encode(username, UTF_8));
        if (found.size() != 1) {
            throw new AssertionError("expected exactly one Keycloak user " + username + ", found " + found.size());
        }
        return found.get(0);
    }

    public static List<String> sessionIds(String userId) {
        List<String> ids = new ArrayList<>();
        admin("/users/" + userId + "/sessions").forEach(session -> ids.add(session.path("id").asText()));
        return ids;
    }

    /** Effective realm roles, composites expanded: what the user actually holds. */
    public static List<String> effectiveRealmRoles(String userId) {
        List<String> names = new ArrayList<>();
        admin("/users/" + userId + "/role-mappings/realm/composite").forEach(role -> names.add(role.path("name").asText()));
        return names;
    }

    public static JsonNode client(String clientId) {
        return admin("/clients?clientId=" + URLEncoder.encode(clientId, UTF_8)).get(0);
    }

    public static void setClientAttribute(String clientId, String attribute, String value) {
        ObjectNode client = (ObjectNode) client(clientId);
        JsonNode attributes = client.path("attributes");
        ObjectNode editable = attributes.isObject() ? (ObjectNode) attributes : client.putObject("attributes");
        editable.put(attribute, value);
        Response response = Http.put(adminUrl("/clients/" + client.path("id").asText()))
                .bearer(masterRealmToken())
                .json(client)
                .send();
        if (response.status() != 204) {
            throw new AssertionError("updating client " + clientId + ": " + response);
        }
    }

    // ----------------------------------------------------------------------------

    public static JsonNode expect(Response response, int status, String what) {
        if (response.status() != status) {
            throw new AssertionError(what + ": expected " + status + ", got " + response);
        }
        return response.body() == null || response.body().isBlank() ? Http.JSON.nullNode() : response.json();
    }

    private static Tokens tokens(JsonNode json) {
        return new Tokens(
                json.path("access_token").asText(null),
                json.path("refresh_token").asText(null),
                json.path("id_token").asText(null));
    }

    private static String adminUrl(String path) {
        return KEYCLOAK + "/admin/realms/" + Platform.setting("KEYCLOAK_REALM", "platform") + path;
    }
}
